chore(docs): common patterns and anti patterns in aztec.nr (#3413)
AztecProtocol/aztec-packages: docs/docs/dev_docs/contracts/resources/common_patterns/main.md

---
title: Common Patterns and Anti-Patterns
---

There are many common patterns have been devised by the Aztec core engineering team and the work of the external community as we build Aztec.nr contracts internally (see some of them [here](https://github.com/AztecProtocol/aztec-packages/tree/master/yarn-project/noir-contracts)).

This doc aims to summarize some of them!

Similarly we have discovered some anti-patterns too (like privacy leakage) that we will point out here!

## Common Patterns
### Safe Math and SafeU120
Field operations may overflow/underflow. Hence we have built a SafeMath library that you can use [based on instructions here](../dependencies.md#safe-math)

> Comparison on Field is also not possible today. For such cases, we recommend using u120, which is wrapped under the SafeU120 class found in the SafeMath library.

Author: can confirm the compiler does error! I will also modify the heading to "comparing Field operations"!

Here you approve someone to transfer funds publicly on your behalf

### Prevent the same user flow from happening twice or avoiding replay attacks using nullifiers 
E.g. you don't want a user to subscribe once they have subscribed already. Or you don't want them to vote twice once they have done that. How do you prevent this?

Emit a nullifier to prevent them from such "replay attacks". This is also why in authwit, we emit a nullifier, to prevent someone from reusing their approval. 

#include_code assert_valid_authwit_public /yarn-project/aztec-nr/authwit/src/auth.nr rust

Note be careful to ensure that the nullifier is not deterministic and that no one could do a preimage analysis attack. More in [the anti pattern section on deterministic nullifiers](#deterministic-nullifiers)

Note - you could also create a note and send it to the user. The problem is there is nothing stopping the user from not presenting this note when they next interact with the function. 

### Reading public storage in private
You can't read public storage in private domain. But nevertheless reading public storage is desirable. There are two ways:

> 1. You pass the data as a parameter to your private method and later assert in public that the data is correct. E.g.:

Reviewer: Should we warn that this leaks information about the private function that was executed (and the data which has been read)? This almost seems like an anti-pattern for privacy?

```rust
struct Storage {
   token: PublicState<Field, 1>,
}

contract Bridge {
    
    #[aztec(private)]
    fn burn_token_private(
        token: AztecAddress, // pass token here since this is a private method but can't access public storage
        amount: Field,
    ) -> Field {
        ...
    #include_code call_assert_token_is_same /yarn-project/noir-contracts/src/contracts/token_bridge_contract/src/main.nr raw
    }
    #include_code assert_token_is_same /yarn-project/noir-contracts/src/contracts/token_bridge_contract/src/main.nr raw
}
```

2. For public storage that changes infrequently, use the slow updates tree! More details TBD

### When calling a public function from private, try to mark the public function as `internal`
This ensures your flow works as intended and that no one can call the public function without going through the private function first!

### Passing data from public to private

> You have public balance and want to move them into the private domain. You shouldn't pass your address that should receive the notes in private, because that leaks privacy (duh!). So what do you do?

Reviewer: These sentences aren't clear to me.

1. You have to create a note in public domain and can't encrypt it, because you can't leak the public key of the receiver. 
2. So how do you control who can claim this note? Pass a hash of a secret instead of the address. And then in the private domain, pass the preimage (the secret) to later claim your funds

So you have to create a custom note in the public domain that is not encrypted by some owner - we call such notes a "TransparentNote" since it is created in public, anyone can see the amount and the note is not encrypted by some owner.

This pattern discussed in detail in [writing a token contract section in the shield() method](../../../tutorials/writing_token_contract.md#shield) and [redeem_shield() method](../../../tutorials/writing_token_contract.md#redeem_shield).

### Discovering my notes

> When you send someone a note, the note hash gets encrypted and added to the merkle tree, but the receiver needs to know they receive a note so they can attempt to decrypt leaves in the tree and add it to their PXE. There are two ways you can discover your notes:

Reviewer 2: ```suggestion
When you send someone a note, the note hash gets added to the [note hash tree](../../../../concepts/advanced/data_structures/trees#note-hash-tree). To spend the note, the receiver needs to get the note itself (the note hash preimage). There are two ways you can get a hold of your notes:
```

> 1. When sending someone a note, use `emit_encrypted_log` (and encrypt the log with their public key). This way as the PXE processes all encrypted logs, it realizes there is a note for itself. [More info here](../../syntax/events.md)

Reviewer 2: ```suggestion
1. When sending someone a note, use `emit_encrypted_log` (the function encrypts the log in such a way that only a recipient can decrypt it). PXE then tries to decrypt all the encrypted logs, and stores the successfully decrypted one. [More info here](../../syntax/events.md)
```

> 2.  Manually using `pxe.addNote()` - What if the contract doesn't emit such logs? Or you creating a note in the public domain and want to consume it in private domain (`emit_encrypted_log` shouldn't be called in the public domain because everything is public), like in the previous section where we created a TransparentNote in public. 

Reviewer 2: > What if the contract doesn't emit such logs?

This is a bit weird as it's up to the developer to set this.

Instead I would describe here that a dev might be inclined here to not emit the logs to save gas.

#include_code pxe_add_note yarn-project/end-to-end/src/e2e_cheat_codes.test.ts typescript

In the token contract, TransparentNotes are stored in a set called "pending_shields" which is in storage slot 5. See [here](../../../tutorials/writing_token_contract.md#contract-storage)

### Randomness in notes
Notes are hashed and stored in the merkle tree. While notes do have a header with a `nonce` field that ensure two exact notes still can be added to the note hash tree (since hashes would be different), preimage analysis can be done to reverse-engineer the contents of the note.

> Hence, it could be a good idea to add a "randomness" field to your note to prevent such attacks. 

Reviewer 2: ```suggestion
Hence, it's necessary to add a "randomness" field to your note to prevent such attacks.
```

There are several patterns here:
1. Give the contract a key and share it amongst all participants. This leaks privacy, as anyone can see all the notes in the contract.
2. `Unshield` funds into the contract - this is used in the [Uniswap tutorial where a user sends private funds into a Uniswap Portal contract which eventually withdraws to L1 to swap on L1 Uniswap](../../../tutorials/uniswap/swap_privately.md). This works like ethereum - to achieve contract composability, you move funds into the public domain. This way the contract doesn't even need keys.

There are several other designs we are discussing through [in this discourse post](https://discourse.aztec.network/t/how-to-handle-private-escrows-between-two-parties/2440) but they need some changes in the protocol or in our demo contract. If you are interested in this discussion, please participate in the discourse post!

## Anti Patterns
There are mistakes one can make to reduce their privacy set and therefore make it trivial to do analysis and link addresses. Some of them are:

### Emitting unencrypted log in private domain

> When emitting unencrypted log in private domain, think twice about the content you put there because the log in unencrypted and therefore anyone can see

Reviewer 2: I would not even mention this here since the function will be eventually removed and now it looks like this:

<img width="1116" alt="image" src="https://github.com/AztecProtocol/aztec-packages/assets/13470840/39ae33d8-64a6-4d26-858b-043963601494">

### Share Private Notes
If you have private state that needs to be handled by more than a single user (but no more than a handful), you can add the note commitment to the private data tree, and then encrypt the note once for each of the users that need to see it. And if any of those users should be able to consume the note, you can generate a random nullifier on creation and store it in the encrypted note, instead of relying on the user secret.

## Anti Patterns
There are mistakes one can make to reduce their privacy set and therefore make it trivial to do analysis and link addresses. Some of them are:

### Passing along your address when calling a public function from private
If you have a private function which calls a public function, remember that sequencer can see any parameters passed to the public function. So try to not pass any parameter that might leak privacy (e.g. `from` address)

PS: when calling from private to public, `msg_sender` is the contract address which is calling the public function.

### Deterministic nullifiers

> In the [avoid replay attacks section](#prevent-the-same-user-flow-from-happening-twice-or-avoiding-replay-attacks-using-nullifiers), we recommended using nullifiers. But what you put in the nullifier is also as important.

Reviewer 2: This link should be updated.
